mongodb-community-operator/pkg/tls/tls.go

```diff
@@ -0,0 +1,113 @@
+package tls
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/mongodb/mongodb-kubernetes/mongodb-community-operator/pkg/kube/secret"
+)
+
+const (
+	CAMountPath             = "/var/lib/tls/ca/"
+	OperatorSecretMountPath = "/var/lib/tls/server/" //nolint
+
+	tlsSecretCertName = "tls.crt"
+	tlsSecretKeyName  = "tls.key"
+	tlsSecretPemName  = "tls.pem"
+)
+
+type TLSConfigurableResource interface {
+	metav1.Object
+	TLSSecretNamespacedName() types.NamespacedName
+	TLSOperatorSecretNamespacedName() types.NamespacedName
+}
+
+// ensureTLSSecret will create or update the operator-managed Secret containing
+// the concatenated certificate and key from the user-provided Secret.
+// Returns the file name of the concatenated certificate and key
+func EnsureTLSSecret(ctx context.Context, getUpdateCreator secret.GetUpdateCreator, resource TLSConfigurableResource) (string, error) {
+	certKey, err := getPemOrConcatenatedCrtAndKey(ctx, getUpdateCreator, resource.TLSSecretNamespacedName())
+	if err != nil {
+		return "", err
+	}
+	// Calculate file name from certificate and key
+	fileName := OperatorSecretFileName(certKey)
+
+	operatorSecret := secret.Builder().
+		SetName(resource.TLSOperatorSecretNamespacedName().Name).
+		SetNamespace(resource.TLSOperatorSecretNamespacedName().Namespace).
+		SetField(fileName, certKey).
+		SetOwnerReferences(resource.GetOwnerReferences()).
+		Build()
+
+	return fileName, secret.CreateOrUpdate(ctx, getUpdateCreator, operatorSecret)
+}
+
+// getCertAndKey will fetch the certificate and key from the user-provided Secret.
+func getCertAndKey(ctx context.Context, getter secret.Getter, secretName types.NamespacedName) string {
+	cert, err := secret.ReadKey(ctx, getter, tlsSecretCertName, secretName)
+	if err != nil {
+		return ""
```

Comment:
is it correct to not error in case referenced secret is not there? If we look for something and it's not there, isn't it better to requeue the reconcile and try again?